qiskit_experiments/curve_analysis/curve_data.py

```diff
@@ -126,6 +128,181 @@ def get_subset_of(self, index: Union[str, int]) -> "CurveData":
+    correl: np.ndarray = dataclasses.field(init=False, repr=False)
+    ufloat_params: Dict[str, uncertainties.UFloat] = dataclasses.field(init=False, repr=False)
+
+    def __post_init__(self):
+        self._set_correlation_matrix(self.covar)
+        self._set_ufloat_params(self.covar)
+
+    @property
+    def x_range(self) -> Tuple[float, float]:
+        """Range of X values."""
+        return min(self.x_data), max(self.x_data)
+
+    @property
+    def y_range(self) -> Tuple[float, float]:
+        """Range of Y values."""
+        return min(self.y_data), max(self.y_data)
+
+    @deprecated_function("0.5", "Use '.ufloat_params' which returns a dictionary instead.")
+    def fitval(self, key: str) -> uncertainties.UFloat:
+        """Deprecated. Return UFloat parameter specified by the key."""
+        return self.ufloat_params[key]
+
+    def _set_ufloat_params(self, covar: np.ndarray):
+        """A helper function to compute ufloat parameters from covariance matrix."""
+        if self.params is None:
+            object.__setattr__(self, "ufloat_params", None)
+            return
+
+        if covar is not None:
+            ufloat_fitvals = uncertainties.correlated_values(
+                nom_values=[self.params[name] for name in self.var_names],
+                covariance_mat=covar,
+                tags=self.var_names,
+            )
+        else:
+            # Invalid covariance matrix. Std dev is set to nan, i.e. not computed.
+            ufloat_fitvals = uarray(
+                nominal_values=[self.params[name] for name in self.var_names],
+                std_devs=np.full(len(self.var_names), np.nan),
+            )
+
+        # Combine fixed params and fitting variables into a single dictionary
+        # Fixed parameter has zero std_dev
+        ufloat_params = {}
+        for name in self.params.keys():
+            try:
+                uind = self.var_names.index(name)
+                ufloat_params[name] = ufloat_fitvals[uind]
+            except ValueError:
+                ufloat_params[name] = uncertainties.ufloat(self.params[name], std_dev=0.0)
+
+        object.__setattr__(self, "ufloat_params", ufloat_params)
+
+    def _set_correlation_matrix(self, covar: np.ndarray):
+        """A helper function to compute correlation matrix from covariance matrix."""
+        if covar is not None:
+            # This is how uncertainties computes correlation matrix
+            stdevs = np.sqrt(np.diag(covar))
+            correl = covar / stdevs / stdevs[:, np.newaxis]
+        else:
+            correl = None
+
+        object.__setattr__(self, "correl", correl)
+
+    def __json_encode__(self):
+        # Remove auto-initialized parameters
+        data_dict = dataclasses.asdict(self)
+        del data_dict["correl"]
+        del data_dict["ufloat_params"]
+
+        return data_dict
+
+    @classmethod
+    def __json_decode__(cls, value):
+        return cls(**value)
+
+    def __copy__(self):
```

Comment:
I'm not sure what is the point of shallow copying a frozen dataclass?

Reply:
This is no longer dataclass, but the point is to explicitly say this is not configurable by users, or developers in any subroutine since modification of result may invalidate the result. However, unfortunately there is force copying in DbAnalysisResult for `.value`. This invalidates correlation of ufloats.